packages/@lwc/engine-core/src/framework/def.ts

```diff
@@ -248,6 +249,10 @@ export function getComponentInternalDef(Ctor: unknown): ComponentDef {
             );
         }
 
+        if (process.env.NODE_ENV !== 'production') {
+            checkVersionMismatch(Ctor);
```

Comment:
What about doing this at component registration when the module evaluates, in `registerComponent`? This would surface the error earlier.